streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorder.java

```diff
@@ -17,117 +17,231 @@
-    private Sensor compactionTimeAvgSensor;
-    private Sensor compactionTimeMinSensor;
-    private Sensor compactionTimeMaxSensor;
     private Sensor numberOfOpenFilesSensor;
     private Sensor numberOfFileErrorsSensor;
 
-    final private Map<String, Statistics> statisticsToRecord = new HashMap<>();
-    final private String metricsScope;
-    final private String storeName;
+    private final Map<String, Statistics> statisticsToRecord = new ConcurrentHashMap<>();
+    private final String metricsScope;
+    private final String storeName;
+
+    private enum State { NEW, RUNNING, NOT_RUNNING, ERROR, MANUAL }
+    private volatile State state = State.NEW;
+    private final Duration recordingInterval = Duration.ofMinutes(10);
+    private final boolean startRecordingThread;
+    private Thread thread;
 
     public RocksDBMetricsRecorder(final String metricsScope, final String storeName) {
+        this(metricsScope, storeName, true);
+    }
+
+    // visible for testing
+    RocksDBMetricsRecorder(final String metricsScope, final String storeName, final boolean startRecordingThread) {
         this.metricsScope = metricsScope;
         this.storeName = storeName;
+        this.startRecordingThread = startRecordingThread;
+        final LogContext logContext = new LogContext(String.format("[RocksDB Metrics Recorder for %s] ", storeName));
+        log = logContext.logger(RocksDBMetricsRecorder.class);
+    }
+
+    public boolean isRunning() {
+        return state == State.RUNNING;
+    }
+
+    public boolean error() {
+        return state == State.ERROR;
+    }
+
+    public void addStatistics(final String storeName,
+                              final Statistics statistics,
+                              final StreamsMetricsImpl streamsMetrics,
+                              final TaskId taskId) {
+        if (state == State.NEW) {
+            initSensors(streamsMetrics, taskId);
+            if (!startRecordingThread) {
+                state = State.MANUAL;
+            }
+        }
+        if (state == State.NEW || state == State.NOT_RUNNING) {
+            createThread();
+        } else {
+            if (statisticsToRecord.containsKey(storeName)) {
+                throw new IllegalStateException("Statistics for store \"" + storeName + "\" has been already added. "
+                    + "This is a bug in Kafka Streams.");
+            }
+        }
+        statistics.setStatsLevel(StatsLevel.EXCEPT_DETAILED_TIMERS);
+        statisticsToRecord.put(storeName, statistics);
+        log.debug("Added statistics for store {}", storeName);
+        if (state == State.NEW || state == State.NOT_RUNNING) {
+            state = State.RUNNING;
+            thread.start();
+        }
+    }
+
+    private void waitForThreadToDie() {
+        if (thread == null) {
+            return;
+        }
+        boolean wait = true;
+        log.debug("Wait for recording thread to die");
+        while (wait) {
```

Comment:
This seems suspicious...
* Shouldn't all blocking operations have timeouts?
* Should we be swallowing and ignoring InterruptedExceptions?

It seems like a recipe for Streams to hang forever un-killably. But I feel like I'm missing something.